fuel-vm/src/checked_transaction.rs

```diff
@@ -0,0 +1,870 @@
+            transaction,
+            metadata,
+            checks_bitmask: Checks::Basic,
+            gas_used_by_predicates: 0,
+        }
+    }
+
+    /// Returns reference on inner transaction.
+    pub fn transaction(&self) -> &Tx {
+        &self.transaction
+    }
+
+    /// Returns the metadata generated during the check for transaction.
+    pub fn metadata(&self) -> &Tx::Metadata {
+        &self.metadata
+    }
+
+    /// Returns the bitmask of all passed checks.
+    pub fn checks(&self) -> &Checks {
+        &self.checks_bitmask
+    }
+
+    /// Gas used by predicate checking. If the predicates have not
+    /// been checked, returns zero.
+    pub fn gas_used_by_predicates(&self) -> Word {
+        self.gas_used_by_predicates
+    }
+
+    /// Performs check of signatures, if not yet done.
+    pub fn check_signatures(mut self) -> Result<Self, CheckError> {
+        if !self.checks_bitmask.contains(Checks::Signatures) {
+            self.transaction.check_signatures()?;
+            self.checks_bitmask.insert(Checks::Signatures);
+        }
+        Ok(self)
+    }
+}
+
+impl<Tx: IntoChecked + ExecutableTransaction + fuel_tx::field::GasLimit> Checked<Tx>
+where
+    Self: Clone,
+    <Tx as IntoChecked>::Metadata: crate::interpreter::CheckedMetadata,
+{
+    /// Check predicates, if not yet done.
+    pub fn check_predicates(mut self, params: ConsensusParameters, gas_costs: GasCosts) -> Result<Self, CheckError> {
+        if !self.checks().contains(Checks::Predicates) {
+            let checked = Interpreter::<PredicateStorage>::check_predicates(self.clone(), params, gas_costs)?;
+            self.checks_bitmask.insert(Checks::Signatures);
+            self.gas_used_by_predicates = checked.gas_used();
+        }
+        Ok(self)
+    }
+}
+
+// #[cfg(feature = "internals")]
+impl<Tx: IntoChecked + Default> Default for Checked<Tx> {
+    fn default() -> Self {
+        Tx::default()
+            .into_checked(Default::default(), &Default::default())
+            .expect("default tx should produce a valid fully checked transaction")
+    }
+}
+
+impl<Tx: IntoChecked> From<Checked<Tx>> for (Tx, Tx::Metadata) {
+    fn from(checked: Checked<Tx>) -> Self {
+        let Checked {
+            transaction, metadata, ..
+        } = checked;
+
+        (transaction, metadata)
+    }
+}
+
+impl<Tx: IntoChecked> AsRef<Tx> for Checked<Tx> {
+    fn as_ref(&self) -> &Tx {
+        &self.transaction
+    }
+}
+
+// #[cfg(feature = "internals")]
+impl<Tx: IntoChecked> AsMut<Tx> for Checked<Tx> {
+    fn as_mut(&mut self) -> &mut Tx {
+        &mut self.transaction
+    }
+}
+
+impl<Tx: IntoChecked> Borrow<Tx> for Checked<Tx> {
+    fn borrow(&self) -> &Tx {
+        self.transaction()
+    }
+}
+
+/// Performs checks for a transaction
+pub trait IntoChecked: Checkable + Sized {
+    /// Metadata produced during the check.
+    type Metadata: Sized;
+
+    /// Returns transaction that passed all `Checks`.
+    /// TODO: Predicate checks
```

Comment:
Are predicate checks going to be added once this is merged into the target branch?

Reply:
Yes